src/test/base32_tests.cpp

```diff
@@ -23,14 +23,14 @@ BOOST_AUTO_TEST_CASE(base32_testvectors)
 
     // Decoding strings with embedded NUL characters should fail
     bool failure;
-    (void)DecodeBase32(std::string("invalid", 7), &failure);
-    BOOST_CHECK_EQUAL(failure, true);
-    (void)DecodeBase32(std::string("AWSX3VPP", 8), &failure);
-    BOOST_CHECK_EQUAL(failure, false);
+    (void)DecodeBase32(std::string("invalid", 8), &failure);
```

Comment:
This fails before and after, indicating that the failure has nothing to do with NUL, no?

Reply:
Good point. I also don't fully understand what any of these tests are trying to achieve? Wouldn't something like
```
diff --git a/src/test/base32_tests.cpp b/src/test/base32_tests.cpp
index 26ed78e2c2..60bb523f42 100644
--- a/src/test/base32_tests.cpp
+++ b/src/test/base32_tests.cpp
@@ -23,13 +23,13 @@ BOOST_AUTO_TEST_CASE(base32_testvectors)
 
     // Decoding strings with embedded NUL characters should fail
     bool failure;
-    (void)DecodeBase32(std::string("invalid", 8), &failure);
-    BOOST_CHECK(failure);
-    (void)DecodeBase32(std::string("AWSX3VPP"), &failure);
+    (void)DecodeBase32(std::string("my======", 8), &failure);
     BOOST_CHECK(!failure);
-    (void)DecodeBase32(std::string("AWSX3VPP\0invalid", 16), &failure);
+    (void)DecodeBase32(std::string("my======\0", 9), &failure);
+    BOOST_CHECK(failure);
+    (void)DecodeBase32(std::string("\0my======", 9), &failure);
     BOOST_CHECK(failure);
-    (void)DecodeBase32(std::string("AWSX3VPPinvalid", 16), &failure);
+    (void)DecodeBase32(std::string("my=\0=====", 10), &failure);
     BOOST_CHECK(failure);
 }
``` 
do more according to what the comment on L24 implies?

Reply:
> no?

no :-)

Before this PR `DecodeBase32(std::string("invalid", 7), &failure)` fails because the supplied input has invalid length (7) and is not padded to a proper length (8) with `=`.

With this PR the same call fails due to the `ValidAsCString()` check early in `DecodeBase32()` (this check passes before this PR). The trailing `\0` upsets the `ValidAsCString()` check which I think was the intent.

Reply:
Sorry, with this PR it is not "the same call", but `DecodeBase32(std::string("invalid", 8), &failure)`.

Reply:
Because the failure reason is not returned, it might be good to use a valid Base32 string, which properly decodes and then make it an invalid cstr, which fails. Otherwise it is harder to guess which failure reason was the intent from just reading the test code.

Reply:
> it might be good to use a valid Base32 string, which properly decodes and then make it an invalid cstr, which fails

The tests below already do that with `AWSX3VPP` (length 8, valid) and `AWSX3VPP\0invalid` (length 16, invalid due to embedded `\0`).

Reply:
To my understanding `AWSX3VPP` is not a valid base32 encoded string, but perhaps I'm missing something obvious?

Reply:
Why not? It has proper length (8) and no embedded `\0`s. `DecodeBase32()` decodes is successfully.

Reply:
Thanks, I see your point and why I misunderstood the test in this case. To me the value `AWXS3VPP` looks like a magic string chosen on purpose and not on random, and I expected it to actually decode into something more human readable, like the other test values in the file on L14-15.

Reply:
Added some comments to make it clearer.